README.md
# ImageSorter
ImageSorter is a GUI-based image processing tool which provides the following functions.
1. Generate sliced images from an input image.
2. Sort image files into separate directories based on annotation information generated via [RectLabel](https://rectlabel.com/).

# Installation
Download the file in thie repository and put "CutSort.app" somewhere you like.
"CutSort.app" is an executable file on Mac OS X 10.15 or later, which was built using Pyinstaller.

# Dependencies
ImageSorter is written in Python and requires libraries listed below.  
- cv2
- scikit-image
- scipy
- numpy
- pandas
- csv
- xml

All codes are only tested using Python3.7.

The executable file of ImageSorter (ImageSorter.app) is also found in this repository.
ImageSorter.app is expected to be able to run on Windows, Mac OS, and Linux machines.

# Usage
Execute ImageSorter.app or type 
```
python /path/to/imagesorter/mian.py
```

When you execute ImageSorter, you will see a window like shown below.  
<kbd>![top window](./images/topwindow.png)</kbd>  

Click the "Image Slicer" tab to slice a image into pieses.  
"File sorter" enable you to classify image files in an input folder into separate folders based on annotation information generated via [RectLabel](https://rectlabel.com/).  



## Image Slicer
<kbd>![slicer](./images/slicer.png)</kbd>  

Image Slicer cuts a large image into pieces with a specified size and overlapps between images.
Please specify an input image via clicking the "Input image file" button.
The width and height of sliced images can be specified in px.

Similarly, "step width of image slicing" and "step height of image slicing" indicate how much a sliced image overlaps along the horizontal axis and the vertical axis with the next image.
If you set values as shown the image above, the first image of slices starts from the top-left corner of the input image with 500 x 500 px.
The next slice of the image will starts at the 251th pixel from the left end of the input image as specified at "Set step width of image slicing".
After the slicing hit at the right end of the input image, the next slice starts from the left end again but at the 251the pixel from the top as specified at "Set step height of image slicing".

All of the sliced images will be created in the directory specified as "Output directory".
If the directory specified at "Output directory" does not exist, it will be created.
The output directory includes an image coordinate file which shows original locations of sliced images in the input image.

## File Sorter
<kbd>![sorter](./images/sorter.png)</kbd>

File Sorter sorts image files into separate directories based on annotations given via [RectLabel](https://rectlabel.com/).
Based on the annotation information in a json file generated by RectLabel, this function create directories and put image files.
The directory containing the input image files to be sorted should be specified as "Input directory", while "Output directory" indicates the directory under which directories will be created to sort the input image files.

File Sorter accepts three situations which input images are:
1. sliced images annotatted via an objecte detection model.
2. single (non-sliced) images annotatted via an objecte detection model.
3. single (non-sliced) images annotatted via an image classification model.

In the situation 1, a same object can be annotated as different classes in separate files of sliced images with overlaps each other.
You can select one best annotation for the object by activating the radio button "Vote to decide annotations".
In that case, "Percentage of overlapped area" need to be specified.
File Sorter algorithm first generate the original iamge from the sliced images with location information stored in an iamge coordinate file and then put all annotation information over the image. If multiple annotations indicated by rectangles are found, File Sorter evaluates whether percentages of overlaping areas covered by the annotation rectangles exceed the value specified at "Percentage of overlapped area". The annotations evaluated as "overlaps" will be merge as one annotation. If the "overlaped" annotations have conflicts, e.g. different annotations for one object, and if "Vote to decide annotaions" is active, File Sorter decides one most frequently annotatted class for the object. Otherwise, it leave two annotations for the object.


News.md
# Update info
- 2022/4/28: Bug fix: Get width and height infomation of a given image, but not rgb coordinate.

- 2021/1/15: Release v1.0


colorcodes.py
#!/usr/bin/python3
# coding: UTF-8


def rgb2hex(r, g, b):
    """RGB値をHTMLなどで使われる16進数表現に変換"""
    return '#%02X%02X%02X' % (r, g, b)


def make_gray_code(i):
    gray_tone = (255, 245, 220, 211, 192, 169, 128, 105, 0)
    if 0 <= i < len(gray_tone):
        gray = gray_tone[i]
        return rgb2hex(gray, gray, gray)
    else:
        return 'Error'


def make_rgb_code(n, i):
    """引数:n=色数(3の倍数)、i=色番号"""

    if n % 3 != 0:
        return 'Error:nが3の倍数では無い'

    def rgb_tone(x):
        return max(abs(256 * 3 // 2 - 256 * 3 * x // n) - 256 // 2, 1) - 1

    r = rgb_tone(i)
    g = rgb_tone((i - n * 1 // 3) % n)
    b = rgb_tone((i - n * 2 // 3) % n)
    return rgb2hex(r, g, b)


def make_color_code(n, i, gray=False):
    """引数:n=色数、i=色番号、gray=False(カラー),True(白黒)"""
    if 0 <= i < n:
        return make_rgb_code(n, i)
    elif gray:
        return make_gray_code(i - n)
    else:
        return 'Error'


if __name__ == '__main__':
    for i in range(12):
        print(make_color_code(12, i, gray=False))  # 色の数=12
    print('-------')
    for i in range(12, 12 + 9):
        print(make_color_code(12, i, gray=True))  # 色の数=12

imageslicer.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Sep  7 10:18:43 2020

@author: ftom
"""

import os
import csv
from skimage import io

def imageslicer(in_file, out_dir, size_x, size_y, step_x, step_y):    
            
    if not os.path.exists(out_dir):
        os.mkdir(out_dir)

    img = io.imread(in_file)
    root, ext = os.path.splitext(in_file)
    basename = os.path.basename(root)
    height, width = img.shape[:2]

    slice_x0 = 0
    slice_y0 = 0
    slice_size_x = size_x
    slice_size_y = size_y
    slice_step_x = step_x
    slice_step_y = step_y

    cord_list = os.path.join(out_dir, "slice_coordinate.csv")  
    with open(cord_list, 'w') as f:
        tmp = ["image_file", "x0", "x1", "y0", "y1"]
        writer = csv.writer(f, lineterminator='\n')
        writer.writerow(tmp)
    
        while True:
            slice_x1 = slice_x0 + slice_size_x
            slice_y1 = slice_y0 + slice_size_y
            
            if slice_x1 > width:
                slice_x1 = width
                slice_x0 = slice_x1 - slice_size_x
            if slice_y1 > height:
                slice_y1 = height
                slice_y0 = slice_y1 - slice_size_y
    
            sliced_img = img[slice_y0:slice_y1, slice_x0:slice_x1]
            out_file = os.path.join(out_dir, basename + "_" + str(slice_y0) + "-" + str(slice_y1) + "_" + str(slice_x0) + "-" + str(slice_x1) + ext)
            io.imsave(out_file, sliced_img)
            
            tmp = [out_file, slice_x0, slice_x1, slice_y0, slice_y1]
            writer = csv.writer(f, lineterminator='\n')
            writer.writerow(tmp)
            
            if slice_x1 == width:
                if slice_y1 == height:
                    break
                slice_x0 = 0
                slice_y0 = slice_y0 + slice_step_y
            else:
                slice_x0 = slice_x0 + slice_step_x


main.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Mon Sep  7 10:53:13 2020

@author: ftom
"""

import os
import tkinter as tk
import tkinter.ttk as ttk
import tkinter.filedialog
from skimage import io
import tkinter.messagebox
import threading as th
import imageslicer as isl
import filesorter as fs


class CutsortGUI(ttk.Frame):
    
    def __init__(self, master=None):
        super().__init__(master)
        self.configure(padding=50)
        self.pack()
        self.initial_menu()
        
    def initial_menu(self):
        # The initial appearance
        s = ttk.Style()
        s.configure('init.TButton', font=('Helvetica', '18'))
        self.filesorter_button = ttk.Button(self,
                                  text = "Image Slicer",
                                  style = 'init.TButton',
                                  command = self.open_slicer)
        self.filesorter_button.pack()
        
        self.imgslicer_button = ttk.Button(self,
                                  text = "File Sorter",
                                  style = 'init.TButton',
                                  command = self.open_sorter)
        self.imgslicer_button.pack(pady=(30, 0))

        
    def back_to_menu(self):
        children = self.winfo_children()
        for child in children:
            child.destroy()
        self.initial_menu()
            
            
    def in_dir_dialog(self):
        iDir = os.path.abspath(os.path.dirname(__file__))
        in_dir = tk.filedialog.askdirectory(initialdir=iDir)
        
        if len(in_dir) == 0:
            self.in_dir.set('Please select an input directory.')
        else:
            self.in_dir.set(in_dir)
            
    def out_dir_dialog(self):
        iDir = os.path.abspath(os.path.dirname(__file__))
        out_dir = tk.filedialog.askdirectory(initialdir=iDir)
        if not os.path.exists(out_dir):
            os.mkdir(out_dir)
        if len(out_dir) == 0:
            self.out_dir.set('Please select an output directory.')
        else:
            self.out_dir.set(out_dir)
            
    def in_file_dialog(self):
        fTyp = [("image file", "*.jpg *.jpeg *.png *.tif *.tiff")]
        iDir = os.path.abspath(os.path.dirname(__file__))
        in_file = tk.filedialog.askopenfilename(filetypes=fTyp, initialdir=iDir)
        if len(in_file) == 0:
            self.in_file.set('Please select an input image file.')
        else:
            self.in_file.set(in_file)
            img_file = self.in_file.get()
            img = io.imread(img_file)
            height, width = img.shape[:2]
            img_h_text = 'Height = ' + str(height) + ' pixels'
            self.img_h.set(img_h_text)
            img_w_text = 'Width = ' + str(width) + ' pixels'
            self.img_w.set(img_w_text)       
        
    def open_slicer(self):
        children = self.winfo_children()
        for child in children:
            child.destroy()
            
        s = ttk.Style()
        s.configure('isl.TButton', font = ('Helvetica', '12', 'bold'))
        s.configure('small.TLabel', font = ('Helvetica', '10'))
        s.configure('regular.TLabel', font = ('Helvetica', '12'))
         
        back_button = ttk.Button(self,
                                 text = "Back to Menu",
                                 style = 'isl.TButton',
                                 command = self.back_to_menu)
        back_button.pack()


        in_file_button = ttk.Button(self,
                                  text = "Input image file",
                                  style = 'isl.TButton',
                                  command = self.in_file_dialog)
        in_file_button.pack(pady=(30, 10))
        
        self.in_file = tk.StringVar()
        self.in_file.set('Please select an input image file.')
        in_file_label = ttk.Label(self,
                                  textvariable = self.in_file,
                                  style = 'small.TLabel')
        in_file_label.pack(pady = (0, 5))
        
        self.img_h = tk.StringVar()
        self.img_h.set('NA')
        img_h_label = ttk.Label(self,
                                textvariable = self.img_h, 
                                style = 'small.TLabel')
        img_h_label.pack(pady = (0, 5))
    
        self.img_w = tk.StringVar()
        self.img_w.set('NA')
        img_w_label = ttk.Label(self,
                                textvariable = self.img_w, 
                                style = 'small.TLabel')
        img_w_label.pack(pady = (0, 10))
    
        size_x_label = ttk.Label(self,
                                 text='Set width of sliced images', 
                                 style = 'regular.TLabel')
        size_x_label.pack(pady = 0)
        self.size_x = ttk.Entry(self)
        self.size_x.insert(0, "500")
        self.size_x.pack(pady = (0, 10))
        
        size_y_label = ttk.Label(self,
                                 text='Set height of sliced images', 
                                 style = 'regular.TLabel')
        size_y_label.pack(pady = 0)
        self.size_y = ttk.Entry(self)
        self.size_y.insert(0, "500")
        self.size_y.pack(pady = (0, 10))
    
        step_x_label = ttk.Label(self,
                               text='Set step width of image slicing', 
                               style = 'regular.TLabel')
        step_x_label.pack(pady = 0)
        self.step_x = ttk.Entry(self)
        self.step_x.insert(0, "250")
        self.step_x.pack(pady = (0, 10))
        
        step_y_label = ttk.Label(self,
                               text='Set step height of image slicing', 
                               style = 'regular.TLabel')
        step_y_label.pack(pady = 0)
        self.step_y = ttk.Entry(self)
        self.step_y.insert(0, "250")
        self.step_y.pack(pady = (0, 10))

        out_dir_button = ttk.Button(self,
                                   text = "Output directory",
                                   style = 'isl.TButton',
                                   command = self.out_dir_dialog)
        out_dir_button.pack(pady=(0, 10))
        
        self.out_dir = tk.StringVar()
        self.out_dir.set ('Please select an output directory.')
        out_dir_label = ttk.Label(self,
                                  textvariable=self.out_dir,
                                 style = 'regular.TLabel')
        out_dir_label.pack(pady = 0)
        
        run_isl = ttk.Button(self,
                               text = "Run",
                               style = 'isl.TButton',
                               command = self.run_slicer)
        run_isl.pack(pady=(20, 0))
            
    def open_sorter(self):
        
        def show_type1():
            setting_frame.pack_forget()
            label1.grid_forget()
            widget1.grid_forget()
            widget2.grid_forget
            in_dir_button.pack_forget()
            in_dir_label.pack_forget()
            in_file_button.pack_forget()
            in_file_label.pack_forget()
            out_dir_button.pack_forget()
            out_dir_label.pack_forget()
            run_button.pack_forget()
            setting_frame.pack(pady = (30, 20))
            label1.grid(row = 0, column = 0)
            widget1.grid(row = 1, column = 0)
            widget2.grid(row = 4, column = 0)
            in_dir_button.pack(pady=(0, 10))
            in_dir_label.pack(pady = (0, 20))
            in_file_button.pack(pady=(0, 10))
            in_file_label.pack(pady = (0, 20))
            out_dir_button.pack(pady=(0, 10))
            out_dir_label.pack(pady = 0)
            run_button.pack(pady=(40,0))
            
        def show_type2():
            setting_frame.pack_forget()
            label1.grid_forget()
            widget1.grid_forget()
            widget2.grid_forget
            in_dir_button.pack_forget()
            in_dir_label.pack_forget()
            in_file_button.pack_forget()
            in_file_label.pack_forget()
            out_dir_button.pack_forget()
            out_dir_label.pack_forget()
            run_button.pack_forget()
            in_dir_button.pack(pady=(0, 10))
            in_dir_label.pack(pady = (0, 20))
            out_dir_button.pack(pady=(0, 10))
            out_dir_label.pack(pady = 0)
            run_button.pack(pady=(40,0))
            
        def show_type3():
            setting_frame.pack_forget()
            label1.grid_forget()
            widget1.grid_forget()
            widget2.grid_forget
            in_dir_button.pack_forget()
            in_dir_label.pack_forget()
            in_file_button.pack_forget()
            in_file_label.pack_forget()
            out_dir_button.pack_forget()
            out_dir_label.pack_forget()
            run_button.pack_forget()
            in_dir_button.pack(pady=(0, 10))
            in_dir_label.pack(pady = (0, 20))
            out_dir_button.pack(pady=(0, 10))
            out_dir_label.pack(pady = 0)
            run_button.pack(pady=(40,0))
                                           
        children = self.winfo_children()
        for child in children:
            child.destroy()
            
        s = ttk.Style()
        s.configure('fs.TButton', font = ('Helvetica', '12', 'bold'))
        s.configure('small.TLabel', font = ('Helvetica', '10'))
        s.configure('regular.TLabel', font = ('Helvetica', '12'))
        s.configure('regular.TLabelframe', font = ('Helvetica', '12'))
        s.configure('regular.TRadiobutton', font = ('Helvetica', '12'))
        s.configure('regular.TEntry', font = ('Helvetica', '12'))
         
        back_button = ttk.Button(self,
                                 text = "Back to Menu",
                                 style = 'fs.TButton',
                                 command = lambda:self.back_to_menu())
        back_button.pack()
        
        rb_frame = ttk.Labelframe(self,
                                  text = 'Input type', 
                                  padding = (10),
                                  style = 'regular.TLabelframe')
        rb_frame.pack(pady = (30, 20))
        self.type_val = tk.IntVar()
        self.type_val.set(1)
        rb1 = ttk.Radiobutton(rb_frame, 
                              text = 'Object detection on sliced images',
                              value = 1,
                              variable = self.type_val,
                              style = 'regular.TRadiobutton',
                              command = show_type1)
        rb2 = ttk.Radiobutton(rb_frame, 
                              text = 'Object detection on single images',
                              value = 2,
                              variable = self.type_val,
                              style = 'regular.TRadiobutton',
                              command = show_type2)
        rb3 = ttk.Radiobutton(rb_frame, 
                              text = 'Image classification',
                              value = 3,
                              variable = self.type_val,
                              style = 'regular.TRadiobutton',
                              command = show_type3)
        rb1.grid(row = 0, column = 0)
        rb2.grid(row = 1, column = 0)
        rb3.grid(row = 2, column = 0)
        
        setting_frame = ttk.Labelframe(self,
                                  text = 'Settings to summarize annotations', 
                                  padding = (10),
                                  style = 'regular.TLabelframe')
        setting_frame.pack(pady = (30, 20))
        self.threshold = tk.StringVar()
        self.voting = tk.BooleanVar()
        self.voting.set(False)
        label1 = ttk.Label(setting_frame, 
                              text = 'Percentage of overlaped area',
                               style = 'regular.TLabel')
        widget1 = ttk.Entry(setting_frame,
                              textvariable = self.threshold,
                              style = 'regular.TEntry')
        widget1.insert(0, "0")
        widget2 = ttk.Checkbutton(setting_frame, 
                              text = 'Vote to decide annotations',
                              variable = self.voting,
                              style = 'regular.TRadiobutton')
        label1.grid(row = 0, column = 0)
        widget1.grid(row = 1, column = 0)
        widget2.grid(row = 4, column = 0)
        
        in_dir_button = ttk.Button(self,
                                  text = "Input directory",
                                  style = 'fs.TButton',
                                  command = self.in_dir_dialog)
        in_dir_button.pack(pady=(0, 10))
        
        self.in_dir = tk.StringVar()
        self.in_dir.set('Please select an input directory.')
        in_dir_label = ttk.Label(self,
                                 textvariable=self.in_dir,
                                 style = 'regular.TLabel')
        in_dir_label.pack(pady = (0, 20))
        
        in_file_button = ttk.Button(self,
                                  text = "Original image file",
                                  style = 'isl.TButton',
                                  command = self.in_file_dialog)
        in_file_button.pack(pady=(0, 10))
        self.in_file = tk.StringVar()
        self.in_file.set('Please select the original image file of sclices.')
        in_file_label = ttk.Label(self,
                                  textvariable = self.in_file,
                                  style = 'regular.TLabel')
        in_file_label.pack(pady = (0, 20))
        
        
        out_dir_button = ttk.Button(self,
                                  text = "Output directory", 
                                  style = 'fs.TButton',
                                  command = self.out_dir_dialog)
        out_dir_button.pack(pady=(0, 10))
        
        self.out_dir = tk.StringVar()
        self.out_dir.set ('Please select an output directory.')
        out_dir_label = ttk.Label(self,
                                  textvariable=self.out_dir, 
                                  style = 'regular.TLabel')
        out_dir_label.pack(pady = 0)
        
        run_button = ttk.Button(self,
                               text = "Run",
                               style = 'fs.TButton',
                               command = self.run_sorter)
        run_button.pack(pady=(40,0))
        
    def run_slicer(self):
        def check_thread():
            if pb.winfo_exists():
                if t_slicer.is_alive():
                    self.after(20, check_thread)
                else:
                    pb_bar.stop()
                    pb.grab_release()
                    pb.destroy()
                    tk.messagebox.showinfo("Image Slicer", "Done!")
        
        pb = tk.Toplevel(self)
        pb.grab_set()
        pb.focus_force()
        pb.title('Running Image Slicer')
        pb_frame = ttk.Labelframe(pb, text = 'Running...')
        pb_frame.configure(padding=30)
        pb_frame.pack()
        pb_bar = ttk.Progressbar(pb_frame, orient='horizontal', length=200, mode='indeterminate')
        pb_bar.pack()
        
        in_file = self.in_file.get()
        out_dir = self.out_dir.get()
        size_x = int(self.size_x.get())
        size_y = int(self.size_y.get())
        step_x = int(self.step_x.get())
        step_y = int(self.step_y.get())
        
        check1 = in_file == 'Please select an input iamge.'
        check2 = out_dir == 'Please select an output directory.'
        if check1:
            tk.messagebox.showinfo("Image Slicer", "Invalid settings.\nPlease specify an input image.")
            pb.destroy()
        elif check2:
            tk.messagebox.showinfo("Image Slicer", "Invalid settings.\nPlease specify an output directory.")
            pb.destroy()
        else:
            t_slicer = th.Thread(target=isl.imageslicer, args=(in_file, out_dir, size_x, size_y, step_x, step_y, ))
            t_slicer.daemon = True
            pb_bar.start(10)
            t_slicer.start()
            self.after(20, check_thread)
            
    def run_sorter(self):
        def check_thread():
            if pb.winfo_exists():
                if t_sorter.is_alive():
                    self.after(20, check_thread)
                else:
                    pb_bar.stop()
                    pb.grab_release()
                    pb.destroy()
                    tk.messagebox.showinfo("File Sorter", "Done!")
                
        pb = tk.Toplevel(self)
        pb.grab_set()
        pb.focus_force()
        pb.title('Running Image Sorter')
        pb_frame = ttk.Labelframe(pb, text = 'Running...')
        pb_frame.configure(padding=30)
        pb_frame.pack()
        pb_bar = ttk.Progressbar(pb_frame, orient='horizontal', length=200, mode='indeterminate')
        pb_bar.pack()
        
        in_dir = self.in_dir.get()
        check1 = in_dir == 'Please select an input directory.'
        out_dir = self.out_dir.get()
        check2 = out_dir == 'Please select an output directory.'
        in_file = self.in_file.get()
        check3 = in_file == 'Please select the original image file of sclices.'
        
        data_type = self.type_val.get()
        check4 = data_type == 1
        threshold = float(self.threshold.get())
        voting = self.voting.get()
        if check1:
            tk.messagebox.showinfo("File Sorter", "Invalid settings.\nPlease specify an input directory.")
            pb.destroy()
        elif check2:
            tk.messagebox.showinfo("File Sorter", "Invalid settings.\nPlease specify an output directory.")
            pb.destroy()
        elif check3 and check4 and (not voting):
            tk.messagebox.showinfo("File Sorter", "Invalid settings.\nPlease specify the original image or turn on voting.")
            pb.destroy()
        else:
            t_sorter = th.Thread(target=fs.filesorter, args=(in_dir, out_dir, data_type, in_file, threshold, voting,))
            t_sorter.daemon = True
            pb_bar.start(10)
            t_sorter.start()
            self.after(20, check_thread)

if __name__ == '__main__':
    global root
    root = tk.Tk()
    root.title("Cut&Sort")
    CutsortGUI(master=root)
    root.mainloop()
